src/Autodesk.Forge.Core/ForgeHandler.cs

```
@@ -176,22 +179,24 @@ protected virtual async Task RefreshTokenAsync(HttpRequestMessage request, bool
                 }
             }
         }
-        protected virtual async Task<(string, TimeSpan)> Get2LeggedTokenAsync(string scope, CancellationToken cancellationToken)
+        protected virtual async Task<(string, TimeSpan)> Get2LeggedTokenAsync(string user, string scope, CancellationToken cancellationToken)
         {
             using (var request = new HttpRequestMessage())
             {
                 var config = this.configuration.Value;
-                if (string.IsNullOrEmpty(config.ClientId))
+                var clientId = string.IsNullOrEmpty(user) ? config.ClientId : config.Agents[user].ClientId;
+                if (string.IsNullOrEmpty(clientId))
                 {
                     throw new ArgumentNullException($"{nameof(ForgeConfiguration)}.{nameof(ForgeConfiguration.ClientId)}");
                 }
-                if (string.IsNullOrEmpty(config.ClientSecret))
+                var clientSecret = string.IsNullOrEmpty(user) ? config.ClientSecret : config.Agents[user].ClientSecret;
```

even if `user` is not null or empty, how can we know config.Agents[user] exists? should we use call `TryGetValue` to avoid exceptions?

---

I think we do want to throw an exception in this case. The programmer made a mistake, and they need to correct their code.